stratum/hal/lib/barefoot/bfrt_node.cc

```diff
@@ -131,6 +138,14 @@ ::util::Status BfrtNode::CommitForwardingPipelineConfig() {
   RETURN_IF_ERROR(bf_sde_interface_->AddDevice(device_id_, bfrt_config_));
 
   // Push pipeline config to the managers.
+  RETURN_IF_ERROR(p4runtime_bfrt_translator_->PushForwardingPipelineConfig(
+      bfrt_config_.programs(0).p4info()));
+
+  // Augment the P4Info so managers will use the original bitwith from P4 code
+  // for every fields (e.g, 9-bit port number instead of 32-bits).
+  ASSIGN_OR_RETURN(const auto& low_level_p4info,
+                   p4runtime_bfrt_translator_->GetLowLevelP4Info());
+  *bfrt_config_.mutable_programs(0)->mutable_p4info() = low_level_p4info;
```

Comment:
This is not a good program flow / API design. In the node we first save the p4info, then push it to a sub-manager, then read it back from it and overwrite to local version.

Reply:
Maybe we can also have the same logic for the translator (save and commit). 
And we store a translated version when we are creating `bfrt_config_`?

https://github.com/stratum/stratum/blob/main/stratum/hal/lib/barefoot/bfrt_node.cc#L109
```diff
- *program->mutable_p4info() = config.p4info();
+ *program->mutable_p4info() = bfrt_p4runtimr_translator_.TranslateP4Info(config.p4info());
```

Reply:
Maybe we should always keep the original version(one form the control plane to the bfrt node) and push the translated version to managers

Reply:
I think there is a way to structure this. Everything above BfrtNode is (potentially) untranslated. In the write path, the node translates all requests to SDK form and passes them on to the managers. Therefore, everything below node is translated. So we can and should pass the low level p4info down to the managers.

Reply:
Looking deeper into the `P4TableMapper`, I think that design is the way to go.

Reply:
> I think there is a way to structure this:
>     * Everything above BfrtNode is (potentially) untranslated.
>     * In the write path, the node translates all requests to SDK form and passes them on to the managers.
>     * Therefore, everything below the node is translated.
>     * So we can and should pass the low level p4info down to the managers.

I think that's what I am currently doing now.
I add a new change to the translator and the bfrt node. Now bfrt node stores the untouched pipeline config.
When committing pipeline config to bfrt mangers, bfrt node will create a copy for the pipeline config and modify it and push to those managers.

I can also add the translator to those managers, the reason I didn't put is simply want to reduce the total number of dependency.

Reply:
> I think that's what I am currently doing now. (unless I missed something?)

no, that's right. It's a summary of the current state.

Reply:
got it.
I am trying to figure out roles for different components, for example

bfrt_node: pass P4Runtime requests to managers that can handle them
bfrt_x_manager: Translate P4Runtime messages to the internal data structure and pass to the SDE interface.
bfrt_p4runtime_translator: Translate P4Runtime message to another P4Runtime message with a different value

And looks like P4TableMapper does both translating the data structure(P4- > common) and maybe the value. And bcm managers translate the common entry to the bcm specific one.

Reply:
I think we can move the translator to those managers to keep the bfrt node clean.
And calls the translation in manager functions when bfrt node pass P4Runtime requests to the manager
so we can keep some logic in the lower level

Reply:
jep, this might work